state/application.go

```diff
@@ -1494,30 +1508,65 @@ func (a *Application) ChangeScale(scaleChange int) (int, error) {
 
 // SetScale sets the application's desired scale value.
 // This is used on CAAS models.
-func (a *Application) SetScale(scale int) error {
+func (a *Application) SetScale(scale int, generation int64, force bool) error {
 	if scale < 0 {
 		return errors.NotValidf("application scale %d", scale)
 	}
+	svcInfo, err := a.ServiceInfo()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	logger.Tracef(
+		"SetScale DesiredScaleProtected %v, DesiredScale %v -> %v, Generation %v -> %v",
+		svcInfo.DesiredScaleProtected(), a.doc.DesiredScale, scale, svcInfo.Generation(), generation,
+	)
+	if svcInfo.DesiredScaleProtected() && !force && scale != a.doc.DesiredScale {
+		return errors.Forbiddenf("SetScale(%d) without force while desired scale %d is not applied yet", scale, a.doc.DesiredScale)
+	}
+	if !force && generation < svcInfo.Generation() {
+		return errors.Forbiddenf(
+			"application generation %d can not be reverted to %d", svcInfo.Generation(), generation,
+		)
+	}
+
 	buildTxn := func(attempt int) ([]txn.Op, error) {
 		if attempt > 0 {
+			if err := a.Refresh(); err != nil {
+				return nil, errors.Trace(err)
+			}
 			alive, err := isAlive(a.st, applicationsC, a.doc.DocID)
 			if err != nil {
 				return nil, errors.Trace(err)
 			} else if !alive {
 				return nil, applicationNotAliveErr
 			}
-			if err := a.Refresh(); err != nil {
-				return nil, errors.Trace(err)
-			}
 		}
-		return []txn.Op{{
+		ops := []txn.Op{{
 			C:  applicationsC,
 			Id: a.doc.DocID,
-			Assert: bson.D{{"life", Alive},
+			Assert: bson.D{
+				{"life", Alive},
 				{"charmurl", a.doc.CharmURL},
-				{"unitcount", a.doc.UnitCount}},
+				{"unitcount", a.doc.UnitCount},
+			},
 			Update: bson.D{{"$set", bson.D{{"scale", scale}}}},
-		}}, nil
+		}}
+		cloudSvcDoc := cloudServiceDoc{
+			DocID: a.globalKey(),
+		}
+		if force {
+			// scale from cli.
+			cloudSvcDoc.DesiredScaleProtected = true
+		} else {
+			// scale from cluster always has a valid generation (>= current generation).
+			cloudSvcDoc.Generation = generation
+		}
+		cloudSvcOp, err := buildCloudServiceOps(a.st, cloudSvcDoc)
```

Comment:
as per above, best just to pass in params and return txns, not have the caller need to know about doc structs

buildCloudServiceOps(a, force, generation)